### PR TITLE
Have readonly constants in the main rear script

### PR DESCRIPTION
**Pull request by jsmeix**
In particular have '/mnt/local' as readonly
variable RECOVERY_FS_ROOT in usr/sbin/rear
see https://github.com/rear/rear/issues/708
and use $RECOVERY_FS_ROOT in the scripts instead
of hardcoded '/mnt/local' but currently only a few scripts
are adapted so that I can test if it works at all (and for me
it does work).

Have several more readonly constants in the main rear script
like SHARE_DIR CONFIG_DIR VAR_DIR LOG_DIR ...
see https://github.com/rear/rear/issues/678
and fixed if readonly constants are modified in other scripts
as far as currently possible for me.

Additionally the DISKLAYOUT_FILE variable is moved from
usr/share/rear/conf/default.conf to usr/sbin/rear because
it is not meant to be configured by the user
see https://github.com/rear/rear/issues/678
    
By the way fixed creation of /media /media/cdrom /media/floppy in
usr/share/rear/restore/SUSE_LINUX/91_create_missing_directories.sh
because those directories ar no longer needed since SLE12.

### CODE REVIEW

**Comment by jsmeix** [comment]
FYI:

My following test shows the remaining usage of readonly constants
in other scripts.

Readonly constants are only used in
<pre>
Makefile
packaging/debian/rules
</pre>

My test and its results:

<pre>
github/rear $ for v in $( grep '^readonly' usr/sbin/rear | cut -d' ' -f2- | cut -d '=' -f1 ) ; do echo "v='$v'" ; find * | xargs grep "[^_]$v=" | grep -v ':[[:space:]]*#' ; echo ; done | grep -v '^usr/sbin/rear'

v='PRODUCT'

v='PROGRAM'

v='VERSION'
Makefile:version := $(shell awk 'BEGIN { FS="=" } /^VERSION=/ { print $$2}' $(rearbin))
Makefile:               -e 's#^VERSION=.*#VERSION=$(distversion)#' \

v='RELEASE_DATE'
Makefile:               -e 's#^RELEASE_DATE=.*#RELEASE_DATE="$(release_date)"#' \

v='STARTTIME'

v='CMD_OPTS'

v='SCRIPT_FILE'

v='REAR_DIR_PREFIX'

v='RECOVERY_FS_ROOT'

v='OPTS'

v='ARGS'

v='DEBUG'

v='DEBUGSCRIPTS'

v='DEBUGSCRIPTS_ARGUMENT'

v='DEBUGSCRIPTS_OPPOSITE_ARGUMENT'

v='SIMULATE'

v='STEPBYSTEP'

v='REAR_LOGFILE'

v='RECOVERY_MODE'

v='SHARE_DIR'
Makefile:               -e 's,^SHARE_DIR=.*,SHARE_DIR="$(datadir)/rear",' \
packaging/debian/rules:         -e 's#^SHARE_DIR=.*#SHARE_DIR="/usr/share/rear"#' \

v='CONFIG_DIR'
Makefile:       sed -i -e 's,^CONFIG_DIR=.*,CONFIG_DIR="$(sysconfdir)/rear",' \
packaging/debian/rules:         -e 's#^CONFIG_DIR=.*#CONFIG_DIR="/etc/rear"#' \

v='VAR_DIR'
Makefile:               -e 's,^VAR_DIR=.*,VAR_DIR="$(localstatedir)/lib/rear",' \
packaging/debian/rules:         -e 's#^VAR_DIR=.*#VAR_DIR="/var/lib/rear"#' \

v='LOG_DIR'

v='KERNEL_VERSION'

v='VERSION_INFO'
</pre>

**Comment by jsmeix** [comment]
For me rear still works with that pull request
on a SLES12-SP1 KVM/Qemu virtual machine
with 2.5TB virtual harddisk using SUSE's special 'gpt_sync_mbr'
and SLES12-SP1 default btrfs and xfs for /home 
where I did
<pre>
rear --debugscripts xv mkbackup
</pre>
and
<pre>
rear --debugscripts xv recover
</pre>

Using
<pre>
grep 'readonly variable' /var/log/rear/rear*.log
</pre>
does not show any usage of a readonly variable
neither for "mkbackup" nor for "recover".

FYI regarding using a readonly variable:
<pre>
# readonly foo=Foo

# readonly | grep foo
declare -r foo="Foo"

# foo=Bar
-bash: foo: readonly variable

# unset foo
-bash: unset: foo: cannot unset: readonly variable
</pre>

**Comment by jsmeix** [comment]
As neded we may add more readonly constants
but for now the issue is fixed.

File: usr/share/rear/restore/SUSE_LINUX/91_create_missing_directories.sh [hunk 1]
```diff
@@ -1,7 +1,23 @@
...
+# and the "SUSE Linux Enterprise Server 12 Release Notes" mention it
+# see https://www.suse.com/releasenotes/x86_64/SUSE-SLES/12/ that reads:
+#   "/run/media/<user_name> is now used as top directory for removable
+#    media mount points. It replaces /media , which is not longer available."
+# Therefore for SLE12 "rear recover" must no longer create them.
+# FIXME: The following test for SLE12 products is intentionally sloppy
+# because I <jsmeix@suse.de> have no better idea how to test for various
+# posible SLE12-based products like "SUSE Linux Enterprise Server 12"
+# "SUSE Linux Enterprise Desktop 12" "SUSE Linux Enterprise Server 12 SP1"
+# "SUSE Linux Enterprise Desktop 12 SP1" "SUSE Linux Enterprise <whatever> <whichever>".
```
Comment (schlomo) [inline usr/share/rear/restore/SUSE_LINUX/91_create_missing_directories.sh]: What about the `OS_MASTER_VERSION` or `OS_VERSION` variables? Won't they help here? After all, for that purpose we put them there.

Comment (jsmeix) [inline usr/share/rear/restore/SUSE_LINUX/91_create_missing_directories.sh]: Using versions was my initial attempt but this does not work for me because I do not know what of the many possible versions are affected (and I don't know what future version values there might be - e.g. openSUSE Leap is 42.1).

Furthermore in general I prefer to test for something "real" and not depend on an (abstract) version number.

In short: The current way is intentional.

Comment (schlomo) [inline usr/share/rear/restore/SUSE_LINUX/91_create_missing_directories.sh]: I meant more like

```
if ((  OS_VERSION < 12 )) ; then
```

You said that with SLES12 we don't need to create the directories.

An alternative would be to have a function called in the subdirs for SLE9, SLE10 and SLE11 and nothing in SUSE_LINUX.

Comment (jsmeix) [inline usr/share/rear/restore/SUSE_LINUX/91_create_missing_directories.sh]: "((  OS_VERSION < 12 ))" was the first thing I tried but
it cannot work because OS_VERSION can be things
like "13.2" for openSUSE 13.2 and then:
<pre>
# echo $OS_VERSION
13.2

# if ((  OS_VERSION < 12 )) ; then echo y ; fi
-bash: ((: 13.2: syntax error: invalid arithmetic operator (error token is ".2")
</pre>

Perhaps an OS_VERSION setting of "13.2" is wrong
and needs to be fixed?
But at least currently such an OS_VERSION is allowed
and also used for Red Hat according to
doc/user-guide/02-getting-started.adoc
<pre>
# rear dump
Relax-and-Recover 1.12.0svn497 / 2011-07-11
Dumping out configuration and system information
System definition:
                                    ARCH = Linux-x86_64
                                      OS = GNU/Linux
                               OS_VENDOR = RedHatEnterpriseServer
                              OS_VERSION = 5.6
</pre>

But even if OS_VERSION would be split into
<pre>
OS_MAJOR_VERSION=13
OS_MINOR_VERSION=2
</pre>
it does not really help me in this particular case because
I simply do not know for what exact openSUSE versions
media/cdrom and media/floppy were used and I do not want
to install each openSUSE version to find out what there
actually is.

As I said:
In general I do not like testing for version numbers.
In general I prefer testing for the real thing.

In this particular case I implemented "testing for the real thing"
in a sloppy and indirect way (I know RFC 1925 6a) but
personally I consider this to be still better than obscure
version number tests and I provide intentionally a longer
explanatory comment to tell the user what the issue is.

What does a plain number tell the user?
Basically nothing.

Finally the whole
usr/share/rear/restore/SUSE_LINUX/91_create_missing_directories.sh
looks broken to me on a first glance because I wonder why
directories in /media are not in the backup?

As far as I see on my SLE11 system /media is not a separated filesystem so that I would expect to have all in /media included
in the backup by default.

In contrast on my SLE12 system /run is a separeted filesystem:
<pre>
# findmnt | grep run  
|-/run  tmpfs  tmpfs  rw,nosuid,nodev,mode=755
</pre>
so that things in /run/media should not be in the backup
by default.

Comment (schlomo) [inline usr/share/rear/restore/SUSE_LINUX/91_create_missing_directories.sh]: I just realized this:

```
root@devidc01 ~ # expr 1.5 \< 2 && echo yes
1
yes
root@devidc01 ~ # expr 1.5 \< 1 && echo yes
0
```

But I can also live with your current solution. Having `OS_VERSION_MAJOR` is  good idea but might not work for all distros. Using expr to compare it numerically might work for more.

Comment (jsmeix) [inline usr/share/rear/restore/SUSE_LINUX/91_create_missing_directories.sh]: I found out what there is on openSUSE 13.1 and 13.2:

Both have /etc/os-release and
both do not have media/cdrom and media/floppy

Therefore current SUSE_LINUX/91_create_missing_directories.sh
works for SLE11, openSUSE 13.1 13.2 and SLE12 and then also for openSUSE Leap 42.1 because that is based on SLE12.
